[Doc Feature] Bot sessions (#14081)
MicrosoftDocs/msteams-docs: msteams-platform/bots/how-to/conversations/agent-sessions.md

Reviewer: This needs a TOC entry, what you are seeing in the rendered preview is cached and outdated. Please insert the entry in the _Rich conversational features_ section, just call it "Sessions"

Author: Updated

Reviewer: When we publish this, is the feature in preview? If so, this needs the standard note at the top.

I also want to start getting more crisp about exactly what _preview_ means where we can. Since this has an associated manifest setting, if it requires the `devPreview` version, let's be explicit about that. Does the user need to be checkbox-opted-in to _Developer preview_? What if they aren't?

> [!NOTE]
> Sessions are distinct from threaded replies in channels. A session is a full, independent conversation context within a 1:1 chat and not a reply chain under a single message.

<!--## How topics improve bot conversations

Traditional one-on-one bot chats maintain a single, continuous message thread. As conversations grow longer and span multiple subjects, it becomes harder for users and bots to maintain context. Topics address this challenge by introducing boundaries between conversations while keeping them accessible within the same chat experience.

When topics are enabled:

- Each topic maintains its own conversation history.

- Users can switch between topics without losing context.

- Bots can initiate new topics to organize workflows or updates.

- Notifications and navigation experiences help users track new activity.

For example, a developer productivity bot might create a separate topic for each pull request. Users can review updates, track failures, and resolve issues without mixing unrelated conversations.
-->
## Enable sessions for your agent

Sessions are an opt-in capability that you enable through your app manifest.  

> After you enable sessions and publish the updated app, users see the sessions experience after they install or upgrade the app.

Reviewer: ```suggestion
Sessions are an opt-in capability that you enable through your app manifest. After you enable sessions and publish the updated app, users see the sessions experience after they install or upgrade the app.
```

> [!IMPORTANT]

> > After your agent opts in to sessions, we recommend keeping the feature enabled. Once sessions are enabled on a chat, there is no way to revert it to the exact state of a regular 1:1 chat. Opting out of sessions is **not supported at GA**. Opt-out support is planned as a fast follow-up after GA.

Reviewer: "Support" needs disambiguation here between "you literally can't do it", and "you can do it but we don't take responsibility for anything it breaks."

Please use _general availability_ instead of GA.

Author: Hello @yogapat,

When a developer enables `supportsSessions: true` and later wants to disable it (sets it to false or removes it from the manifest):

- Is it technically impossible: meaning the platform has no mechanism to revert a sessions-enabled chat back to a regular 1:1 chat?

- Or is it technically possible but unsupported: meaning the developer can change the manifest back, but the behavior is undefined/untested and things may break?

cc: @nickwalkmsft

Reviewer: Just flagging that this comment is still open.

Once this is clarified, I anticipate that the first sentence of the alert box should be dropped.

Update your app manifest and set the `supportsSessions` property to `true`.

```json
{
  "bots": [
    {
      "botId": "{{BOT_ID}}",
      "supportsSessions": true
    }
  ]
}
```

After you update the manifest, package and republish your app through the [Developer Portal for Teams](https://dev.teams.microsoft.com/) or [Teams admin center](https://admin.teams.microsoft.com/). Users see the sessions experience after they install or upgrade the app.

> Agents that don't enable sessions continue to use the legacy single chat experience. When sessions are enabled for an existing agent, Teams automatically converts the existing chat history into a default session. No action is required and users don't lose any conversation history.

Reviewer: ```suggestion
Agents that don't enable sessions continue to use the single chat experience. When sessions are enabled for an existing agent, Teams automatically converts the existing chat history into a default session. No action is required and users don't lose any conversation history.
```

Don't call it legacy unless our best practices state that new agents should always use it.

## Send and receive messages in sessions

> Session participation is automatic and requires no special implementation. Your existing message-handling code works without changes.

Reviewer: ```suggestion
For agents that support sessions, participation is automatic and requires no special implementation in code. Your existing message-handling code works without changes.
```

When sessions are enabled, incoming activities include a session-scoped conversation ID. The conversation ID represents the current session and must be treated as an opaque value. Your agent shouldn't parse or construct the conversation ID manually.

Use the conversation ID from the incoming activity when responding.

::: zone pivot="teams-sdk-csharp"

```csharp
var conversationId = context.Activity.Conversation.Id;
await app.Send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-typescript"

```typescript
const conversationId = activity.conversation.id;
await app.send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-python"

```python
conversation_id = ctx.activity.conversation.id
await app.send(conversation_id, "Hello from Bot")
```

::: zone-end

> Responding inside a session works the same way as responding in a normal one-on-one chat. When your agent replies using the conversation ID from the incoming activity, Teams automatically delivers the message to the correct session. Each session maintains independent conversation context.

Reviewer: Move this above the code snippets, and use it to replace "Use the conversation ID from the incoming activity when responding."

> Without sessions, all messages in a 1:1 chat share a single, static `conversationId`. With sessions enabled, each session gets its own unique `conversationId`. The value is always an opaque, encrypted string. Store it and pass it back to the API. If your agent previously cached a `conversationId` from before the user opted into sessions, that cached ID still works and routes messages to the default session.

Reviewer: ```suggestion
If your agent previously cached a `conversationId` from before the user opted into sessions, that cached ID still works and routes messages to the default session.
```

Authorization: Bearer {bot-token}
Content-Type: application/json

{
  "type": "message",
  "text": "Follow-up message in this session"
}
```

#### Reply to a specific activity

```http
POST {serviceUrl}/v3/conversations/{sessionConversationId}/activities
Authorization: Bearer {bot-token}
Content-Type: application/json

{
  "type": "message",
  "text": "Reply to your question",
  "replyToId": "1749012345678"
}
```

#### Get members

```http
GET {serviceUrl}/v3/conversations/{sessionConversationId}/members
Authorization: Bearer {bot-token}
```

> [!NOTE]
> Member operations (`GetMembers`, `GetMember`, `GetProfile`) work the same regardless of which session conversationId you use. They always resolve to the underlying 1:1 chat.

## Create sessions proactively

When sessions are enabled, the existing create conversation operation applied to a 1:1 conversation creates a new session. Two conditions must be met:

1. **The agent must be installed for the user.** If the agent is not installed, session creation is skipped and a regular 1:1 conversation is created instead (no error returned).
2. **The agent's manifest must declare `supportsSessions: true`.**

> Both conditions are checked server-side. An agent that declares `supportsSessions: true` but is not installed for the target user will silently fall back to a regular 1:1 conversation. If both conditions are met, calling the create conversation API with exactly one member and one message activity creates a new session.

Reviewer: ```suggestion
To create a new session with a user from a session-enabled agent, use the create conversation operator with a single message activity, specifying the user as the sole member.
```

I'm not convinced that the "fall back to a 1:1 conversation" behavior is relevant to mention for a user who has not even installed the agent.

> [!IMPORTANT]

> > Each call to the create conversation API creates a new session. There is no API to send a proactive message into an existing session. To send follow-up messages to an existing session, store the conversation ID returned from the initial creation and use it for subsequent replies.

Reviewer: ```suggestion
```

Not needed here. Include a standalone paragraph sentence (not an alert block) about proactively messaging into an existing session in _Send and receive messages in sessions_, because there's already good content there covering the concept that everything important about a session is already encapsulated in the conversation ID.

> For more information, see [proactive messaging](send-proactive-messages.md#create-the-conversation).

Reviewer: ```suggestion
For more information about proactive messaging, see [proactive messaging](send-proactive-messages.md#create-the-conversation).
```

Move this below the snippets.

// Send a follow-up into the same session.
await app.Send(
    sessionConversationId,
    "This message is part of the session.",
    serviceUrl: serviceUrl,
    cancellationToken: cancellationToken);
```

::: zone-end

::: zone pivot="teams-sdk-typescript"

```typescript
// Including exactly one initial `activity` is required for creating a session
// It becomes the first message of the new session.
const resource = await app.api.conversations.create({
  members: [{ id: userId, role: 'user', name: '' }],
  tenantId,
  activity: { type: 'message', text: 'Hello! Starting a new session.' },
});

// resource.id is the session conversationId
const sessionConversationId = resource.id;

await app.send(sessionConversationId, 'This message is part of the session.');
```

::: zone-end

::: zone pivot="teams-sdk-python"

[TODO]

::: zone-end

> [!NOTE]
>
> The Teams SDK calls the same `POST /v3/conversations` endpoint as regular 1:1 creation. No special flag is needed. The server creates a session automatically when the bot is session-enabled and installed.
> The `activity` field is **required** to have exactly one message activity, which becomes the first message of the new session.

> > The returned id (`resource.Id` in C#, `resource.id` in TypeScript) is the session-specific conversationId. Use it for all follow-up messages in that session (via `app.Send` / `app.send`).

Reviewer: ```suggestion
```

Not needed.

## Error codes

Error codes that bots may encounter during session operations:

| HTTP Status | Error Code | Description |
| --- | --- | --- |
| 400 | `BadArgument` | The 1:1 thread already uses a threading mode that cannot be converted to sessions. |
| 400 | `BadSyntax` | Missing or empty request body, or null activity. Session creation requires exactly one activity. |
| 400 | `MissingProperty` | Required fields missing from the request. |
| 401 | `AuthorizationError` | Bot token invalid or expired. |
| 403 | `NotEnoughPermissions` | Bot doesn't have permission to message this user. |
| 403 | `BotDisabledByAdmin` | Tenant admin has disabled the bot. |
| 403 | `BotNotInConversationRoster` | Bot is not installed in the conversation. |
| 404 | `ConversationNotFound` | The conversation thread doesn't exist. |
| 429 | `Throttled` | Rate limit exceeded. Retry after the indicated delay. |
| 500 | `ServiceError` | Unexpected server error. |
| 502 | — | Upstream service timed out while verifying bot session support. |

> ## Best practices

Reviewer: ```suggestion
## Best practices and design guidance
```

### When to enable sessions

Enable sessions when your agent handles multiple distinct tasks or topics with the same user, such as managing separate support tickets, pull requests, or project workflows. Sessions are also a good fit for AI-powered agents, where shorter, focused conversation contexts improve the quality of generated responses. If your agent sends proactive notifications alongside interactive conversations, sessions help keep updates separate from ongoing tasks.

Agents that serve a single, continuous purpose with no need for task separation can continue without sessions.

### Organize sessions around user tasks

Create sessions that represent clear goals or workflows. For example, you might create separate sessions for individual tickets, pull requests, or customer requests.

Clear session separation helps users quickly locate and resume conversations.

### Limit creating too many sessions

> Create a new session only when it represents a distinct task or workflow that benefits from its own conversation context. Avoid creating sessions for transient updates or single-message interactions. Too many sessions make it harder for users to find relevant conversations in the sessions panel. If your agent sends updates based on activity outside of Teams, such as CI/CD results, service alerts, or external system notifications, route them to a single dedicated notifications session instead of creating a new session for each update. Store the conversation ID of the notifications session after you create it, and reuse that ID for subsequent messages.

Reviewer: ```suggestion
Create a new session only when it represents a distinct task or workflow that benefits from its own conversation context. Avoid creating sessions for transient updates or single-message interactions. Too many sessions make it harder for users to find relevant conversations in the sessions panel.

If your agent sends updates based on activity outside of Teams, such as CI/CD results, service alerts, or external system notifications, route them to a single dedicated notifications session instead of creating a new session for each update. Store the conversation ID of the notifications session after you create it, and reuse that ID for subsequent messages.
```

description: Sessions enable agents in Microsoft Teams to organize one-on-one conversations into separate, focused chats, helping maintain context, improve response relevance, and support multi-task workflows.
ms.author: vikasalmal
ms.localizationpriority: high
ms.topic: article
ms.date: 07/10/2026
zone_pivot_groups: teams-sdk-languages
---

<!-- markdownlint-disable MD024 -->

# Manage multiple user conversations with sessions

By default, one-on-one chat between a user and an agent takes place in a single long-running conversation. With sessions, an agent can hold multiple independent one-on-one conversations with a user.

Sessions offer users and agents a structured way to manage multiple tasks or workflows, similar to other modern AI assistant experiences. Encouraging users to organize their interactions into shorter, more focused contexts can also improve the quality of LLM-generated responses.

Sessions is an optional feature that must be enabled for an agent in its manifest configuration.

## User experience

When chatting one-on-one with an agent that supports sessions, controls in the header enable users to create and switch between sessions.

> Clicking on the new session icon starts a new session.​

Reviewer: ```suggestion
Clicking on the new session icon starts a new session.​ The first message in a session becomes the session title.
```

Author: Updated

# [Desktop](#tab/desktop)

:::image type="content" source="../../../assets/images/bots/create-new-session-desktop.png" alt-text="Screenshot shows creating a new session button on desktop" lightbox="../../../assets/images/bots/create-new-session-desktop.png" border="false":::

# [Mobile](#tab/mobile)

:::image type="content" source="../../../assets/images/bots/create-new-session-mobile.png" alt-text="Screenshot shows creating a new session button on mobile" lightbox="../../../assets/images/bots/create-new-session-mobile.png" border="false":::

---

> After a message is sent, the session is created and saved in the sessions panel.

Reviewer: ```suggestion
After a message is sent, the session is created and saved in the sessions panel. The latest message in the session appears as the preview.
```

Author: Updated

- Topics panel displays all active topics and highlights unread topics.

- Activity feed shows notifications such as mentions or reactions within a topic.

- Search helps users locate messages across topics.

Selecting any notification automatically opens the correct topic and message.
-->
With sessions enabled, users get the following capabilities within the one-on-one agent chat:

- **Create sessions**: Start a new session from the chat header to begin a focused conversation.
- **Switch between sessions**: Move between active sessions through the sessions panel without losing context.
- **Track unread activity**: Sessions with new messages are highlighted and badged so users can identify pending updates.
- **Navigate to specific activity**: Select a notification or session entry to jump directly to the relevant message.
- **Access from personal app**: Open the agent through the personal app entry point. Sessions are fully supported in the personal app experience, including deep links.
- **Search across sessions**: Find messages across all sessions using Teams search.

All existing one-on-one agent chat capabilities continue to work within sessions.

> The first message in a session becomes the session title, and the latest message appears as the preview in the sessions panel. Sessions created proactively by the agent appear as unread for the user. Agents can also proactively create new sessions with users to organize notifications, updates, or task-specific conversations.

Reviewer: ```suggestion
Agents can proactively create new sessions with users to organize notifications, updates, or task-specific conversations. Sessions created proactively by an agent appear as unread for the user.

All existing one-on-one agent chat capabilities continue to work within sessions.
```

Author: Updated

When topics are enabled:

- Each topic maintains its own conversation history.

- Users can switch between topics without losing context.

- Bots can initiate new topics to organize workflows or updates.

- Notifications and navigation experiences help users track new activity.

For example, a developer productivity bot might create a separate topic for each pull request. Users can review updates, track failures, and resolve issues without mixing unrelated conversations.
-->
## Enable sessions for your agent

Sessions are an opt-in capability that you enable through your app manifest. After you enable sessions and publish the updated app, users see the sessions experience after they install or upgrade the app.

> [!IMPORTANT]
> After your agent opts in to sessions, we recommend keeping the feature enabled. Once sessions are enabled on a chat, there is no way to revert it to the exact state of a regular 1:1 chat. Opting out of sessions is **not supported at GA**. Opt-out support is planned as a fast follow-up after GA.

> Update your app manifest and set the `supportsSessions` property to `true`.

Reviewer: ```suggestion
To enable sessions for an agent, set the `supportsSessions` property in its app manifest to `true`.
```

Author: Updated

```json
{
  "bots": [
    {
      "botId": "{{BOT_ID}}",
      "supportsSessions": true
    }
  ]
}
```

After you update the manifest, package and republish your app through the [Developer Portal for Teams](https://dev.teams.microsoft.com/) or [Teams admin center](https://admin.teams.microsoft.com/). Users see the sessions experience after they install or upgrade the app.

Agents that don't enable sessions continue to use the single chat experience. When sessions are enabled for an existing agent, Teams automatically converts the existing chat history into a default session. No action is required and users don't lose any conversation history.

## Send and receive messages in sessions

For agents that support sessions, participation is automatic and requires no special implementation in code. Your existing message-handling code works without changes.

When sessions are enabled, incoming activities include a session-scoped conversation ID. The conversation ID represents the current session and must be treated as an opaque value. Your agent shouldn't parse or construct the conversation ID manually.

> Without sessions, all messages in a 1:1 chat share a single, static `conversationId`. With sessions enabled, each session gets its own unique `conversationId`. The value is always an opaque, encrypted string. Store it and pass it back to the API. If your agent previously cached a `conversationId` from before the user opted into sessions, that cached ID still works and routes messages to the default session.

Reviewer: ```suggestion
For agents with sessions enabled, the conversation ID of every received one-on-one chat message activity includes an embedded session identifier. Sending a message using that activity's context or conversation ID will route it to the same session.

This makes agent participation in sessions effectively automatic in most cases, including proactive messaging using the conversation ID. In the example below, the resulting message will be posted to the appropriate session if sessions are enabled, and sent to the long-running one-on-one chat with the user if not.

```

Author: Updated

::: zone pivot="teams-sdk-csharp"

```csharp
var conversationId = context.Activity.Conversation.Id;
await app.Send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-typescript"

```typescript
const conversationId = activity.conversation.id;
await app.send(conversationId, "Hello from Bot");
```

::: zone-end

::: zone pivot="teams-sdk-python"

```python
conversation_id = ctx.activity.conversation.id
await app.send(conversation_id, "Hello from Bot")
```

::: zone-end

> Responding inside a session works the same way as responding in a normal one-on-one chat. When your agent replies using the conversation ID from the incoming activity, Teams automatically delivers the message to the correct session. Each session maintains independent conversation context. To proactively send a message into an existing session, store the session's conversation ID and use the standard proactive messaging pattern.
> 
> If your agent previously cached a `conversationId` from before the user opted into sessions, that cached ID still works and routes messages to the default session.

Reviewer: ```suggestion
If your agent cached a conversation ID prior to enabling sessions, that cached ID still works and sends messages to the default session.

Conversation IDs should be treated as opaque values. Constructing or modifying conversation IDs is not supported.
```

Reviewer: This originally said "If your agent previously cached a conversation ID from before **the user** opted into sessions" (emphasis mine).

Was that a mistake, or does the **user** need to opt in? There's nothing about that in the rest of the article. In either case, there should be a sentence about it in the User Experience section.

Now that I'm recognizing this, I'm seeing one or two other places where we need to make it clear when we say "opt in" that we're talking about the agent implementation, not the user; I'll comment on those.

## Error codes

Error codes that bots may encounter during session operations:

| HTTP Status | Error Code | Description |
| --- | --- | --- |
| 400 | `BadArgument` | The 1:1 thread already uses a threading mode that cannot be converted to sessions. |
| 400 | `BadSyntax` | Missing or empty request body, or null activity. Session creation requires exactly one activity. |
| 400 | `MissingProperty` | Required fields missing from the request. |
| 401 | `AuthorizationError` | Bot token invalid or expired. |
| 403 | `NotEnoughPermissions` | Bot doesn't have permission to message this user. |
| 403 | `BotDisabledByAdmin` | Tenant admin has disabled the bot. |
| 403 | `BotNotInConversationRoster` | Bot is not installed in the conversation. |
| 404 | `ConversationNotFound` | The conversation thread doesn't exist. |
| 429 | `Throttled` | Rate limit exceeded. Retry after the indicated delay. |
| 500 | `ServiceError` | Unexpected server error. |
| 502 | — | Upstream service timed out while verifying bot session support. |

## Best practices and design guidance

### When to enable sessions

> Enable sessions when your agent handles multiple distinct tasks or topics with the same user, such as managing separate support tickets, pull requests, or project workflows. Sessions are also a good fit for AI-powered agents, where shorter, focused conversation contexts improve the quality of generated responses. If your agent sends proactive notifications alongside interactive conversations, sessions help keep updates separate from ongoing tasks.

Reviewer: This should lead with as blanket of a statement as we're comfortable with (i.e. "Most agents should enable sessions"), and should be reworded to assume that basically all agents are "AI-powered".

If "Most agents should enable sessions" is in fact the case, this should be reworked to lean hard on that and instead call out scenarios where it's *not* appropriate.

Author: Updated
